File: first-arch-java/src/main/java/Handler_log.java
import com.google.gson.Gson;
import com.google.gson.JsonElement;
import io.nats.client.Connection;
import io.nats.client.Nats;
import io.nats.client.Subscription;
import io.nats.client.Message;
import io.nats.client.impl.NatsMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import java.io.FileWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeoutException;

import static java.lang.System.exit;

public class Handler_log {
    private static final Logger logger = LoggerFactory.getLogger(Handler_acl.class);


    private static void handler() throws IOException, InterruptedException, TimeoutException {
        Injector injector = new Injector();
        Service aclService = injector.getServiceById("log");
        String topic = aclService.getServiceTopic();
        Map<String, String> message = new HashMap<>();
        message.put("timestamp", "2024-11-28T16:05:34");
        message.put("message", "ciao");
        message.put("severity", "info");
        String result = invokeFunction(topic, message);
        System.out.println("logging result: " + result);
    }

    private static String invokeFunction(String topic, Map<String, String> message) throws IOException, InterruptedException, TimeoutException {
        Connection nc = Nats.connect();
        long start = System.currentTimeMillis();
        Gson g = new Gson();
        String msg = g.toJson(message);
        Message response = nc.request(NatsMessage.builder()
                        .subject(topic)
                        .data(msg)
                        .build(),
                Duration.ofSeconds(5));
        long end = System.currentTimeMillis();
        logger.info("logging function executed in {} ms", end - start);

        String result = new String(response.getData(), StandardCharsets.UTF_8);
        return result;
    }

    public static void main(String[] args) throws IOException, InterruptedException, TimeoutException {
        handler();
        exit(0);
    }
}

File: first-arch-js/handler_log.js
const Injector = require("./Injector.js")
const { connect } = require('nats');
const winston = require("winston");
const logger = winston.createLogger({
    level: "info",
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.json()
    ),
    transports: [
      new winston.transports.File({ filename: "logs.txt" }),
    ],
  });

async function hanlder(){

const injector = new Injector()
const log_service = await injector.getServiceById("log")
const topic = log_service.ServiceTopic
console.log ("topic: ", topic)
const message = {timestamp: "2024-11-28T16:05:34", message: "ciao", severity: "info"}
const result = await invokeFunction(topic, message)

console.log("logging result:", result)

}



async function invokeFunction ( topic, message){
    try {
        // Connect to the NATS server
        const nc = await connect({ servers: "nats://localhost:4222" }); // Replace with your NATS server address
        
        console.log("Connected to NATS");
        const start = Date.now()

        // Publish the message
        const response = await nc.request(topic, JSON.stringify(message));
        const end = Date.now()
        logger.info(`logging function executed in ${end-start} ms `)

        
        // Close the connection
        nc.drain();
        console.log("Connection closed");
        return (response.data.toString())
      } catch (err) {
        console.error("Error publishing message:", err);
      }
}

hanlder()